equational_theories/ForMathlib/GroupTheory/FreeGroup/ReducedWords.lean

```diff
@@ -0,0 +1,269 @@
+  intro red step
+  induction step with
+  | @not l r x b =>
+    unfold reduced at red
+    simp at red
+
+theorem not_step_reduced_iff : reduced L₁ ↔ ∀ L₂, ¬ Step L₁ L₂ := by
+  constructor
+  · exact fun h L₂ => not_step_reduced h
+  · intro h
+    induction L₁ with
+    | nil => exact reduced_nil
+    | cons x tail ih =>
+      cases tail with
+      | nil => exact reduced_singleton
+      | cons y tail =>
+        rw [reduced_cons]
+        constructor
+        · intro ⟨eq1, eq2⟩
+          obtain ⟨x1, x2⟩ := x
+          obtain ⟨y1, y2⟩ := y
+          simp only at eq1 eq2
+          apply h tail
+          rw [eq1, ← eq2]
+          apply Step.cons_not
+        · apply ih
+          intro L₂ step
+          apply h (x :: L₂)
+          exact Step.cons step
+
+theorem reduced_infix : reduced L₂ → L₁ <:+: L₂ → reduced L₁ := Chain'.infix
+
+theorem reduced_min (h : reduced L₁) : Red L₁ L₂ ↔ L₂ = L₁ :=
+  Relation.reflTransGen_iff_eq fun _ => not_step_reduced h
+
+def cyclicallyReduced (L : List (α × Bool)) : Prop := reduced L ∧ ∀ a ∈ L.getLast?, ∀ b ∈ L.head?,
+¬(a.1 = b.1 ∧ (!a.2) = b.2)
+
+@[simp]
+theorem cyclicallyReduced_nil {α} : cyclicallyReduced ([] : List (α × Bool)) := by
+  simp [cyclicallyReduced]
+
+@[simp]
+theorem cyclicallyReduced_singleton {x : (α × Bool )} : cyclicallyReduced [x] := by
+  simp [cyclicallyReduced]
+
+theorem cyclicallyReduced_iff : cyclicallyReduced L ↔ reduced L ∧ ∀ a ∈ L.getLast?, ∀ b ∈ L.head?,
+¬(a.1 = b.1 ∧ (!a.2) = b.2) := by rfl
+
+theorem cyclicallyReduced_cons_append {a b} : cyclicallyReduced (b :: L ++ [a]) ↔ reduced (b :: L ++ [a]) ∧ ¬(a.1 = b.1 ∧ (!a.2) = b.2) := by
+  rw [cyclicallyReduced_iff,List.getLast?_concat]
+  simp
+
+
+theorem reduced_of_cyclicallyReduced (L : List (α × Bool)) : cyclicallyReduced L → reduced L := fun
+  h => h.1
+
+theorem cyclicallyReduced_flatten_replicate (n : ℕ) (L : List (α × Bool)) (h : cyclicallyReduced L):
+cyclicallyReduced (List.replicate n L).flatten := by match n, L with
+| 0, _ => simp
+| n+1, [] => simp
+| n+1, (head::tail) =>
+  unfold cyclicallyReduced at *
+  unfold reduced at *
+  rw [List.chain'_flatten (by simp)]
+  constructor
+  · constructor
+    · simp only [List.mem_replicate, ne_eq, AddLeftCancelMonoid.add_eq_zero, one_ne_zero, and_false,
+      not_false_eq_true, true_and, Bool.not_eq_eq_eq_not, not_and, Bool.not_eq_not, forall_eq] at *
+      apply h.1
+    · apply List.chain'_replicate_of_rel _ h.2
+  · intro a ha b hb
+    simp only [Option.mem_def] at ha hb
+    rw [List.getLast?_flatten_replicate (h := by simp_arith)] at ha
+    rw [List.head?_flatten_replicate (h := by simp_arith)] at hb
+    apply h.2 _ ha _ hb
+
+variable [DecidableEq α]
+
+theorem reduced_iff_eq_reduce : reduced L ↔ reduce L = L := by
+  constructor
+  · intro h
+    rw [← reduced_min h]
+    exact reduce.red
+  · intro h
+    unfold reduced
+    rw [List.chain'_iff_all_append_cons_cons]
+    intro x y l₁ l₂ eq1 eq2
+    obtain ⟨x1, x2⟩ := x
```

Comment:
Can you use lower indices please? Also, you can combine this line with the `intro` above.